src/aleph/schemas/pending_messages.py

```diff
@@ -0,0 +1,175 @@
+"""
+Schemas to process raw messages coming from users to the Aleph network.
+These schemas are used to parse messages coming from the network into
+more practical Python objects.
+
+While extremely similar to the functionalities of the aleph message module
+(of which we reuse some classes), this implementation differs in several
+ways:
+1. We do not expect the `content` key to be provided. At best, we get
+   an `item_content` field for inline type messages. Otherwise,
+   the content has to be fetched (and validated) later from the network.
+2. We do not care for confirmations, as the message we are checking is
+   not even integrated yet.
+
+TODO: this module should reasonably be part of aleph message, if only
+      to make the schemas available for the validation of client data
+      in aleph-client.
+"""
+
+
+import json
+from hashlib import sha256
+from typing import Any, Literal, Optional, Union
+
+from aleph_message.models import (
+    AggregateContent,
+    BaseContent,
+    ForgetContent,
+    PostContent,
+    ProgramContent,
+    StoreContent,
+    HashType,
+)
+from aleph_message.models import MessageType, ItemType
+from pydantic import BaseModel, root_validator, validator
+
+from aleph.exceptions import InvalidMessageError
+from aleph.utils import item_type_from_hash
+
+MAX_INLINE_SIZE = 200000  # 200kb max inline content size.
+
+
+class BasePendingMessage(BaseModel):
+    """
+    A raw Aleph message, as sent by users to the Aleph network.
+    """
+
+    sender: str
+    chain: str
+    signature: str
+    type: MessageType
+    item_content: Optional[str]
+    item_type: ItemType
+    item_hash: str
+    time: float
+    channel: Optional[str] = None
+    content: Optional[BaseContent] = None
+
+    @root_validator(pre=True)
+    def load_content(cls, values):
+        """
+        Preload inline content. We let the CCN populate this field later
+        on for ipfs and storage item types.
+        """
+
+        item_type = values["item_type"]
+        item_content = values.get("item_content")
+        if item_type == ItemType.inline:
+            if item_content is None:
+                raise ValueError("Item content not specified for inline item type")
+
+            if len(item_content) > MAX_INLINE_SIZE:
+                raise ValueError("Message too long")
+            try:
+                values["content"] = json.loads(item_content)
+            except json.JSONDecodeError as e:
+                raise ValueError("Could not load item content") from e
+        else:
+            if item_content is not None:
+                raise ValueError(f"{item_type} messages cannot define item_content")
+
+        return values
+
+    @root_validator()
```

Comment:
Yes, if only because we're accessing other fields. The main problem is that `check_item_hash` uses `item_type` and `check_item_type` uses `item_hash`, so at least one of the functions has to be a root validator.